File: sprint12/JMS/FW/BOJ_2263.cpp
#include <iostream>

using namespace std;

int n;
int inorder[100001], postorder[100001], idx[100001];

void Solution(int inStart, int inEnd, int postStart, int postEnd) {
	if (inStart > inEnd || postStart > postEnd) return;

	int rootIdx = idx[postorder[postEnd]];
	int leftSize = rootIdx - inStart;
	
	cout << inorder[rootIdx] << ' ';

	Solution(inStart, rootIdx - 1, postStart, postStart + leftSize - 1);
	Solution(rootIdx + 1, inEnd, postStart + leftSize, postEnd - 1);
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n;
	for (int i = 1; i <= n; i++) {
		cin >> inorder[i];
		idx[inorder[i]] = i;
	}
	for (int i = 1; i <= n; i++) cin >> postorder[i];

	Solution(1, n, 1, n);
}

File: sprint12/KMS/FW/BOJ_2263.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(100001)
N = int(input())
inorder = list(map(int, input().split()))
postorder = list(map(int, input().split()))
instart = 0
inend = N-1
poststart = 0
postend = N-1
position = [0] * (N+1)
for i in range(N):
    position[inorder[i]] = i


def solution(instart, inend, poststart, postend):
    if (instart > inend) or (poststart > postend):
        return
    root = postorder[postend]

    root_index = position[root]
    left_cnt = root_index-instart
    right_cnt = inend - root_index

    print(root, end=' ')
    solution(instart, instart + left_cnt - 1,
             poststart, poststart + left_cnt - 1)

    solution(inend-right_cnt + 1, inend, postend-right_cnt, postend-1)


solution(instart, inend, poststart, postend)

# 7
# 4 2 5 1 6 3 7
# 4 5 2 6 7 3 1
